File: backend/api/month.go
package api

import (
	"math"
	"net/http"
	"time"

	"github.com/labstack/echo/v4"
	models "github.com/mannx/Bluebook/models"
	"github.com/rs/zerolog/log"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

/*
 * Contains all functions that handle incoming api requests
 *
 */

// endOfWeek provides data totaling the previous week of sales
type endOfWeek struct {
	NetSales          float64
	CustomerCount     int
	ThirdPartyPercent float64
	ThirdPartyTotal   float64
}

// dayViewData is an expanded DayData object with additional properties
type dayViewData struct {
	models.DayData

	ThirdPartyDollar  float64
	ThirdPartyPercent float64
	GrossSales        float64
	DayOfMonth        int       // 1-31 for what day of the month it is
	DayOfWeek         string    // user friendly name of what day it is
	IsEndOfWeek       bool      // is this a tuesday?
	EOW               endOfWeek // end of week data if required
	Tags              []string  // list of tags on this day
	TagID             []uint
	SalesLastWeek     int  // 0 if same, -1 if less, 1 if > than last weeks sales for this day
	Exists            bool // true if found in db, false if auto filled

	// below is the extracted date.  Issues parsing date client side in js so this workaround is used for now
	Day   int
	Month time.Month
	Year  int

	Hockey models.HockeySchedule
}

//
//	Returns data for a given month
//	query url: /API_URL?month=MM&year=YYYY
//

// calcWeekEnding retrieves data for a given week ending and generates the totals
func calcWeekEnding(d time.Time, db *gorm.DB, eow *endOfWeek) {
	// end of week, pull in the required data
	pw := d.Add(-time.Hour * 24 * 6) // get previous 7 days
	dat := make([]models.DayData, 7)

	r := db.Find(&dat, "Date >= ? AND Date <= ?", pw, d)
	if r.Error != nil {
		log.Error().Err(r.Error).Msg("Unable to retrieve data to compute end of week calculations")
	} else {
		net := 0.0
		cc := 0
		tps := 0.0 // 3rd party sales
		gsw := 0.0 // gross sales for hte week

		for _, n := range dat {
			net += n.NetSales
			cc += n.CustomerCount
			tps += n.DoorDash + n.SkipTheDishes
			gsw += n.NetSales + n.BottleDeposit + n.HST
		}

		eow.NetSales = net
		eow.CustomerCount = cc
		eow.ThirdPartyTotal = tps

		if gsw > 0 {
			eow.ThirdPartyPercent = (tps / gsw) * 100.0
			if math.IsNaN(eow.ThirdPartyPercent) {
				log.Warn().Msg("[month.go] 3rd party % is NaN, returning 0")
				eow.ThirdPartyPercent = 0.
			}
		}
	}
}

// GetMonthViewHandler handles returning data for viewing a given month
func GetMonthViewHandler(c echo.Context, db *gorm.DB) error {
	// monthlyView holds the monthly day data along with several other bits of info
	type monthlyView struct {
		Data      []dayViewData
		MonthName string // month in a user friendly format
		Month     int    // the month we are returning data for
		Year      int    // the year we are returning data for
	}
	var month, year int

	err := echo.QueryParamsBinder(c).
		Int("month", &month).
		Int("year", &year).
		BindError()
	if err != nil {
		return err
	}

	// get the ending day for the month
	xmonth := month
	if month == 12 {
		// wrap around
		xmonth = xmonth + 1
	}

	endDay := time.Date(year, time.Month(xmonth+1), 0, 0, 0, 0, 0, time.UTC).Day()

	// retrieve the objects in the given range
	start := time.Date(year, time.Month(month), 1, 0, 0, 0, 0, time.UTC)
	end := time.Date(year, time.Month(month), endDay, 0, 0, 0, 0, time.UTC)
	data := make([]models.DayData, endDay)

	res := db.Order("Date").Find(&data, "Date >= ? AND Date <= ?", start, end)

	if res.Error != nil {
		return LogAndReturnError(c, "Error retrieving data", res.Error)
	}

	//	we no longer pre fill the database with empty rows, instead if a day is missing we generate a new empty node
	//	and only save it to the db if required (a comment is added before any import data)

	mvd := make([]dayViewData, endDay) // get a full month allocated

	if len(data) == 0 {
		// empty month, return a full set of empty dates
		genEmptyMonth(&mvd, start, endDay)
	}

	// keep a running total of days processed, and the data itself.  reset once we hit a week ending
	// if we have less than 7 days when the loop is done, calculate the weekly total on the partial data and
	// add empty rows to pad out the data until a week end is required
	dayCount := 0
	dayList := make([]models.DayData, 0)

	for _, o := range data {
		d := time.Time(o.Date)

		// compute the gross sales (net+hst+bot dep)
		gs := o.NetSales + o.HST + o.BottleDeposit

		eow := endOfWeek{} // initialize the end of week of it is required
		if d.Weekday() == time.Tuesday {
			calcWeekEnding(d, db, &eow)

			// reset day count and holding data
			dayCount = 0
			dayList = make([]models.DayData, 0)
		} else {
			// increment day count and store the current day value
			dayCount += 1
			dayList = append(dayList, o)
		}

		// calculate the average sales for the day based on past 4 weeks
		// we no longer cache this value
		o.WeeklyAverage = calculateWeeklyAverage(d, 4, db)

		tags, ids := GetTags(o.ID, db)

		// caluclate the % of 3rd party sales
		tp := o.DoorDash + o.SkipTheDishes
		tpp := 0.0

		if gs > 0.0 {
			tpp = (tp / gs) * 100.0
			if math.IsNaN(tpp) {
				log.Warn().Msg("[month.go] tpp is NaN, changing to 0")
				tpp = 0.0
			}
		}

		slw := 0
		if o.NetSales > o.WeeklyAverage {
			slw = 1
		} else if o.NetSales < o.WeeklyAverage {
			slw = -1
		}

		// // retrieve related hockey data for this day
		hd, _ := genHockeyData(db, time.Time(o.Date))

		dvd := dayViewData{
			DayData:           o,
			ThirdPartyDollar:  tp,
			ThirdPartyPercent: tpp,
			GrossSales:        gs,
			DayOfMonth:        d.Day(),
			DayOfWeek:         d.Weekday().String(),
			IsEndOfWeek:       d.Weekday() == time.Tuesday,
			EOW:               eow,
			Tags:              tags,
			TagID:             ids,
			SalesLastWeek:     slw,
			Exists:            true,
			Day:               d.Day(),
			Month:             d.Month(),
			Year:              d.Year(),
			Hockey:            hd,
		}

		mvd[d.Day()-1] = dvd
	}

	// if dayCount == 0, we had a perfect week
	if dayCount < 7 && dayCount != 0 {

		// from the first entry in dayList add 6 days to get the normal end of the week
		// if the end of the week is after the end of the month (ie. week ends in next month), we are done
		// ( we don't calculate partial weeks that end in the following month)
		// otherwise call calcWeekEnding with the generated end of week date
		start := time.Time(dayList[0].Date)
		end := start.AddDate(0, 0, 6)

		if start.Month() == end.Month() {
			// calculate week ending and generate a blank entry for the last day of the week
			eow := endOfWeek{}
			calcWeekEnding(end, db, &eow)

			day := models.DayData{
				Date: datatypes.Date(end),
			}

			dvd := dayViewData{
				DayData:     day,
				IsEndOfWeek: true,
				EOW:         eow,
				DayOfMonth:  end.Day(),
				DayOfWeek:   end.Weekday().String(),
				Exists:      true,
			}

			mvd[end.Day()-1] = dvd
		}
	}

	// any missing entries (Exists==false) generate a default node
	for i, n := range mvd {
		if !n.Exists {
			date := start.AddDate(0, 0, i)
			mvd[i] = genEmptyDVD(date)

			hd, _ := genHockeyData(db, date)
			mvd[i].Hockey = hd
		}
	}

	mv := monthlyView{
		Data:      mvd,
		MonthName: time.Month(month).String(),
		Month:     month, // include both month and year in our data
		Year:      year,
	}
	return c.JSON(http.StatusOK, &mv)
}

// returns a list of tags and their id's in seperate lists
func GetTags(id uint, db *gorm.DB) ([]string, []uint) {
	// retrieve all tags for this day
	tags := make([]models.TagData, 0)
	res := db.Find(&tags, "DayID = ?", id)
	if res.Error != nil {
		log.Error().Err(res.Error).Msg("Unable to retrieve tag data")
		return nil, nil
	}

	var tstr []string
	var tids []uint

	for _, obj := range tags {
		var t models.TagList
		e := db.Find(&t, "ID = ?", obj.TagID)
		if e.Error != nil {
			log.Error().Err(e.Error).Msg("Unable to retrieve tag data")
			continue
		}

		tstr = append(tstr, t.Tag)
		tids = append(tids, t.ID) // also save the id of the tag for easier searching
	}

	return tstr, tids
}

// computes the average of a number of weeks starting at the given date
func calculateWeeklyAverage(date time.Time, numWeeks int, db *gorm.DB) float64 {
	var dates []time.Time

	for i := 0; i < numWeeks; i++ {
		// generate the previous date for the given week1
		d := date.AddDate(0, 0, -7*(i+1))
		dates = append(dates, d)
	}

	var total float64
	for _, d := range dates {
		var obj models.DayData
		res := db.Find(&obj, "Date = ?", d)
		if res.Error != nil {
			log.Error().Err(res.Error).Msgf("Unable to retrieve data for: %v", time.Time(d).String())
			continue
		}

		// ignore days with 0 sales or days not found
		if res.RowsAffected == 0 || obj.NetSales == 0 {
			numWeeks -= 1
		}

		total = total + obj.NetSales
	}

	n := total / float64(numWeeks)
	if math.IsNaN(n) {
		log.Warn().Msg("[month.go] weekly avarage is NaN, returning 0")
		return 0.0
	}

	return n
}

func genEmptyMonth(data *[]dayViewData, start time.Time, numDays int) {
	for i := 0; i < numDays; i++ {
		d := start.AddDate(0, 0, i)
		*data = append(*data, genEmptyDVD(d))
	}
}

func genEmptyDVD(date time.Time) dayViewData {
	dvd := dayViewData{
		DayData: models.DayData{
			Date: datatypes.Date(date),
		},
		DayOfWeek:   date.Weekday().String(),
		DayOfMonth:  date.Day(),
		IsEndOfWeek: date.Weekday() == 2,
		Day:         date.Day(),
		Month:       date.Month(),
		Year:        date.Year(),
	}

	return dvd
}

func genHockeyData(db *gorm.DB, date time.Time) (models.HockeySchedule, error) {
	// retrieve related hockey data for this day
	var hockey []models.HockeySchedule
	res := db.Where("Date = ?", date).Find(&hockey)
	if res.Error != nil {
		log.Error().Err(res.Error).Msgf("Unable to retrieve hockey schedule for date: %v", date)
		return models.HockeySchedule{
			Valid: false,
		}, res.Error
	}

	hd := models.HockeySchedule{}

	if res.RowsAffected != 0 {
		hd = hockey[0]
	} else {
		// no data found for this day
		return models.HockeySchedule{Valid: false}, nil
	}

	hd.Valid = true
	// hd.HomeGame = hd.Home == HomeTeamName
	htn := GetHomeTeamName(db)
	hd.HomeGame = hd.Home == htn

	return hd, nil
}

File: backend-rust/src/api/month.rs
#![allow(non_snake_case)]
use chrono::{Datelike, Days, NaiveDate};
use diesel::prelude::*;
use diesel::SqliteConnection;
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;

use crate::api::get_days_in_month;
use crate::api::tags::{get_tags, Tags};
use crate::api::DbError;
use crate::models::day_data::DayData;
use crate::models::hockey::HockeySchedule;

#[derive(Serialize, Deserialize)]
struct EndOfWeek {
    NetSales: i32,
    CustomerCount: i32,
    ThirdPartyTotal: i32,
    GrossSales: i32,
}

#[derive(Serialize, Deserialize)]
pub struct MonthData {
    Data: DayData,
    ThirdPartyDollar: i32,
    GrossSales: i32,
    DayOfWeek: String, // text form of the day of the week (Mon,Tue,etc)
    EndOfWeek: Option<EndOfWeek>,
    Tags: Vec<Tags>,

    // How last 4 week average compares to this day.
    // 1 -> Greater sales this week
    // 0 -> same sales
    // -1 -> Sales less than last week
    SalesLastWeek: i32,
    WeeklyAverage: i32, // 4 week average of this day

    // split out as DayData.DayDate gets serialized as [year, day_in_year] with no month
    Day: u8,
    Month: u8,
    Year: i32,
    Hockey: Option<HockeySchedule>,
}

impl MonthData {
    fn new(data: &DayData) -> Self {
        let tpd = data.SkipTheDishes + data.DoorDash + data.UberEats;

        Self {
            Data: data.clone(),
            ThirdPartyDollar: tpd,
            GrossSales: data.NetSales + data.Hst + data.BottleDeposit + data.Tips,
            DayOfWeek: data.DayDate.weekday().to_string(),
            EndOfWeek: None,
            Tags: Vec::new(),
            SalesLastWeek: 0,
            WeeklyAverage: 0,
            Day: data.DayDate.day() as u8,
            Month: data.DayDate.month() as u8,
            Year: data.DayDate.year(),
            Hockey: None,
        }
    }
}

pub fn get_month_data(
    conn: &mut SqliteConnection,
    month: u32,
    year: i32,
) -> Result<Vec<MonthData>, DbError> {
    use crate::schema::day_data::dsl::*;

    let days = get_days_in_month(year, month);
    let start_day = NaiveDate::from_ymd_opt(year, month, 1).expect("invalid date provided");
    let end_day = NaiveDate::from_ymd_opt(year, month, days).expect("invalid date provided (2)");

    // retrieve the data from the db
    let results = day_data
        .filter(DayDate.ge(start_day).and(DayDate.le(end_day)))
        .filter(Updated.eq(false))
        .order(DayDate)
        .select(DayData::as_select())
        .load(conn)?;

    // build the output month data and the extra calculations we need to do
    let mut data = Vec::new();

    for r in &results {
        let mut md = MonthData::new(r);

        // calculate various fields
        // calculate the week ending information if required
        if md.Data.DayDate.weekday() == chrono::Weekday::Tue {
            md.EndOfWeek = Some(calculate_week_ending(conn, &md.Data)?);
        }

        md.WeeklyAverage = calculate_weekly_average(conn, r.DayDate)?;
        let net_sales = r.NetSales;

        md.SalesLastWeek = match net_sales.cmp(&md.WeeklyAverage) {
            Ordering::Greater => 1,
            Ordering::Equal => 0,
            Ordering::Less => -1,
        };

        // get any tags
        md.Tags = get_tags(conn, r)?;

        // get hockey schedule information if any
        md.Hockey = get_hockey_data(conn, r.DayDate).ok();

        data.push(md);
    }

    // do we have missing days to pad out?
    if data.len() != days as usize {
        let missing = days as usize - data.len();
        let date = match data.last() {
            None => {
                // no entries for the month, set date to end of previous month
                NaiveDate::from_ymd_opt(2025, 1, 1).unwrap()
            }
            Some(date) => date.Data.DayDate,
        };

        for i in 0..missing {
            let new_date = date.checked_add_days(Days::new((i + 1) as u64)).unwrap();
            data.push(MonthData::new(&DayData::new(new_date)));
        }
    }

    Ok(data)
}

fn calculate_week_ending(
    conn: &mut SqliteConnection,
    data: &DayData,
) -> Result<EndOfWeek, DbError> {
    use crate::schema::day_data::dsl::*;

    // get the previous 7 days
    let start_date = data
        .DayDate
        .checked_sub_days(Days::new(6))
        .expect("invalid date provided");

    let results: Vec<DayData> = day_data
        .filter(
            DayDate
                .ge(start_date)
                .and(DayDate.le(data.DayDate))
                .and(Updated.eq(false)),
        )
        .select(DayData::as_select())
        .load(conn)?;

    let gross = results
        .iter()
        .map(|x| x.NetSales + x.Hst + x.BottleDeposit)
        .sum::<i32>();
    let net = results.iter().map(|x| x.NetSales).sum::<i32>();
    let tpt = results
        .iter()
        .map(|x| x.SkipTheDishes + x.DoorDash + x.UberEats)
        .sum::<i32>();
    let customer_count = results.iter().map(|x| x.CustomerCount).sum::<i32>();

    Ok(EndOfWeek {
        NetSales: net,
        ThirdPartyTotal: tpt,
        CustomerCount: customer_count,
        GrossSales: gross,
    })
}

/// Calculates the weekly average of the past 4 weeks to compare to this one
/// TODO: have number of weeks a changable setting?
fn calculate_weekly_average(conn: &mut SqliteConnection, date: NaiveDate) -> Result<i32, DbError> {
    use crate::schema::day_data::dsl::*;

    let start_date = date
        .checked_sub_days(Days::new(4 * 7))
        .expect("invalid start date"); // 4 weeks

    let results: Vec<DayData> = day_data
        .filter(DayDate.ge(start_date).and(DayDate.lt(date)))
        .order(DayDate)
        .select(DayData::as_select())
        .load(conn)?;

    let mut total = 0;
    for r in &results {
        // ignore all but the same day
        if r.DayDate.weekday() == date.weekday() {
            total += r.NetSales;
        }
    }

    Ok(total / 4)
}

fn get_hockey_data(
    conn: &mut SqliteConnection,
    date: NaiveDate,
) -> Result<HockeySchedule, DbError> {
    use crate::schema::hockey_schedule::dsl::*;

    let result = hockey_schedule
        .filter(DayDate.eq(date))
        .first::<HockeySchedule>(conn)?;

    Ok(result)
}
